qubit-sdk/src/main/java/com/qubit/android/sdk/internal/placement/interactor/PlacementQueryAttributesBuilder.kt

```diff
@@ -0,0 +1,97 @@
+package com.qubit.android.sdk.internal.placement.interactor
+
+import com.google.gson.JsonArray
+import com.google.gson.JsonElement
+import com.google.gson.JsonObject
+import com.google.gson.JsonPrimitive
+import com.qubit.android.sdk.internal.placement.repository.PlacementAttributesRepository
+
+internal class PlacementQueryAttributesBuilder(
+    private val placementAttributesRepository: PlacementAttributesRepository
+) {
+
+  companion object {
+    internal const val VISITOR_ATTRIBUTE_KEY = "visitor"
+    internal const val USER_ATTRIBUTE_KEY = "user"
+    internal const val VIEW_ATTRIBUTE_KEY = "view"
+  }
+
+  internal fun buildJson(
+      deviceId: String,
+      customAttributes: JsonObject?
+  ): JsonObject {
+    val attributesJson = JsonObject()
+
+    // 1 visitor attributes should be set by SDK
+    attributesJson.add(VISITOR_ATTRIBUTE_KEY, buildVisitorAttributesJson(deviceId))
+
+    // 2 custom attributes passed by SDK user
+    if (customAttributes != null) {
+      customAttributes.remove(VISITOR_ATTRIBUTE_KEY)
+      for (key in customAttributes.keySet()) {
+        // user/view attributes should contain all fields from schema and nothing more
+        val customAttribute = prepareCustomAttribute(key, customAttributes.get(key))
+        attributesJson.add(key, customAttribute)
+      }
+    }
+
+    // 3 cached attributes - skipped if set by SDK user
+    val cachedAttributes = placementAttributesRepository.load()
+    for (key in cachedAttributes.keys) {
+      if (!attributesJson.has(key)) {
+        attributesJson.add(key, cachedAttributes[key])
+      }
+    }
+
+    // 4 if user/view attributes are missing, then we should add empty ones
+    if (!attributesJson.has(USER_ATTRIBUTE_KEY)) {
+      attributesJson.add(USER_ATTRIBUTE_KEY, buildEmptyUserAttribute())
+    }
+    if (!attributesJson.has(VIEW_ATTRIBUTE_KEY)) {
+      attributesJson.add(VIEW_ATTRIBUTE_KEY, buildEmptyViewAttribute())
+    }
+    return attributesJson
+  }
```

Comment:
Whole this algorithm is complicated IMO. 
There are 2 attributes only. And 3 steps of processing. I every step we execute loop over keys.

For single attribute processing can be described by:
```
  private fun JsonObject.add(
          key: String, customAttributes: JsonObject?,
          cachedAttributes: Map<String, JsonObject>,
          converter: (JsonObject)->JsonObject, emptyBuilder: ()-> JsonObject
  ) {
    customAttributes?.get(key)
            ?.let { jsonElement -> if (jsonElement.isJsonObject) jsonElement.asJsonObject else null }
            // 2 custom attributes passed by SDK user
            ?.let { converter(it) }
            // 3 cached attributes - skipped if set by SDK user
            ?: cachedAttributes[key]
            // 4 if user/view attributes are missing, then we should add empty ones
            ?: emptyBuilder()
                    .let { add(key, it) }
  }
```

Then, whole algorithm can look like this:
```
    val cachedAttributes = placementAttributesRepository.load()

    return JsonObject().apply {
      add(VISITOR_ATTRIBUTE_KEY, buildVisitorAttributesJson(deviceId))

      add(USER_ATTRIBUTE_KEY, customAttributes, cachedAttributes,
              { v -> convertTodUserAttribute(v) }, { buildEmptyUserAttribute() })

      add(VIEW_ATTRIBUTE_KEY, customAttributes, cachedAttributes,
              { v -> convertToViewAttribute(v) }, { buildEmptyViewAttribute() })
    }

```

Reply:
In my solution, I forgot about values in customAttributes other than user/view.

Reply:
Good suggestion. I followed your proposal modifying it slightly and adding custom attributes full support.